grouper/service_account.py

```diff
@@ -199,51 +196,60 @@ def enable_service_account(session, actor, service_account, owner):
 
 
 def service_account_permissions(session, service_account):
-    # type: (Session, ServiceAccount) -> List[ServiceAccountPermission]
-    """Return the permissions of a service account."""
-    permissions = session.query(Permission, ServiceAccountPermissionMap).filter(
+    # type: (Session, ServiceAccount) -> List[ServiceAccountPermissionGrant]
+    """Return the permissions of a service account, including mapping IDs.
+
+    This is used to display the permission grants on a service account page, which has to generate
+    revocation links, so return ServiceAccountPermissionGrant objects that include the mapping ID.
+    """
+    grants = session.query(
+        Permission.name,
+        ServiceAccountPermissionMap.argument,
+        ServiceAccountPermissionMap.granted_on,
+        ServiceAccountPermissionMap.id,
+    ).filter(
         Permission.id == ServiceAccountPermissionMap.permission_id,
         ServiceAccountPermissionMap.service_account_id == service_account.id,
         ServiceAccountPermissionMap.service_account_id == ServiceAccount.id,
         ServiceAccount.user_id == User.id,
         User.enabled == True,
     )
     out = []
-    for permission in permissions:
+    for grant in grants:
         out.append(
-            ServiceAccountPermission(
-                permission=permission[0].name,
-                argument=permission[1].argument,
-                granted_on=permission[1].granted_on,
-                mapping_id=permission[1].id,
+            ServiceAccountPermissionGrant(
+                service_account=service_account,
```

Comment:
Is there a way this is automatically getting the username from `service_account`, or should it be `service_account.user.username`? If the latter I don't see how mypy isn't yelling unless it calls [`__repr__`](https://github.com/dropbox/merou/blob/4f986f60c7ce8c00b5717a890f3890eeb4b5c087/grouper/models/service_account.py#L43)

Reply:
Oh, good catch! And I think the SQLAlchemy model objects aren't fully typed; I've seen it miss other things like that.